### PR TITLE
waiting mode checks in ISIS direct inelastic reduction

### LINKED ISSUE #29765: ISIS Mantid inelastic reduction script fails in waiting for file mode
**Original reporter:** [ISIS]/[Excitation Group], confirmed [ISIS]/[@abuts]

### Expected behavior
ISIS excitation group users run data reduction scripts in waiting mode. Script regularly scans the file system, and, if the requested data file is found, script loads it up and runs reduction process. Only specified file extensions are processed in this mode, i.e. normally files with extension nxs, while raw files which appear on the data search path are ignored. 

### Actual behavior
Users complain that the reduction scripts randomly fails, it looks like the reduction picks up either file with wrong extension (raw) or tries to load nxs files before monitor information has been written to disk. 

### Steps to reproduce the behavior
difficult to reproduce, script continues if the reduction retried. Probably one needs to try to run the reduction when the data writing to the archive is not fully completed.  But users regularly hit this problem during experiments. 

### Platforms affected
Any, but reported on Linux. 

### Probable solution:
Suggested by Freddy Akkerroyd:
"Could you monitor the file \\isis\inst$\NDXMERLIN\Instrument\logs\lastrun.txt - it is only updated when the run has been fully copied to the archive, and contains the number of the last run written. The best approach is only to read the file when its “last modified” date has changed, that way you don’t conflict with it being updated, though as robocopy uses retries it should still be fine if you did read it every time (so long as you open and close the file each time to avoid it being locked for too long)."

The code, monitoring and parsing the log file, should be added to the appropriate part of the reduction script which verifies file presence.

### PR DESCRIPTION
**Pull request by abuts**
**Description of work.**

Fixes #29765 
Its unclear why the reduction was failing in the first place. 
Here I have added check for archive copying log. 

**To test:**

Probably just check if all new and old tests are passing. Any issues may arise from undefined file system state, and the work should add some clarity.

 

*This does not require release notes* because **it is specfic ISIS reduction workflow issue and the people who are affected will be aware anyway**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by abuts** [comment]
@FreddieAkeroyd This modifications should check the archive copying log and implement the algorithm, suggested on 25/09/2020
```
Hi,
 

Could you monitor the file \\isis\inst$\NDXMERLIN\Instrument\logs\lastrun.txt - it is only updated when the run has been fully copied to the archive, and contains the number of the last run written. The best approach is only to read the file when its “last modified” date has changed, that way you don’t conflict with it being updated, though as robocopy uses retries it should still be fine if you did read it every time (so long as you open and close the file each time to avoid it being locked for too long).

 

Regards,

 
Freddie   
 

From: Adroja, Devashibhai (STFC,RAL,ISIS) <devashibhai.adroja@stfc.ac.uk>
Sent: 25 September 2020 12:15
To: Akeroyd, Freddie (STFC,RAL,ISIS) <freddie.akeroyd@stfc.ac.uk>; Mantid Help <mantid-help@mantidproject.org>; Gigg, Martyn (Tessella,RAL,ISIS) <martyn.gigg@stfc.ac.uk>
Cc: Buts, Alex (STFC,RAL,ISIS) <alex.buts@stfc.ac.uk>; Voneshen, David (STFC,RAL,ISIS) <david.voneshen@stfc.ac.uk>
Subject: MERLIN analysis script error

 

Dear Freddie  and Mantid Help.

We run MERLIN analysis script and it waits for new nxs files to appears. However, it seems it cannot find file with extension “.nxs” (for the current run, which is running), but it finds “.raw”, which is not complete (see error below). Hence, script fails. Any suggestion, how to solve this problem? One solution is not to write “.raw” files.

Thanks

With regards

Adroja
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 1]
```diff
@@ -30,6 +31,18 @@ class NonIDF_Properties(object):
           "information" : (4,lambda msg:   logger.information(msg)),
           "debug" :       (5,lambda msg:   logger.debug(msg))}
 
+    # The default location of the archive load log file
+    if platform == "linux" or platform == "linux2":
+        # linux default log file location
+        arhive_upload_log_template = '/archive/NDX{0}/Instrument/logs/lastrun.txt'
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:37]:
```suggestion
        archive_upload_log_template = '/archive/NDX{0}/Instrument/logs/lastrun.txt'
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 2]
```diff
@@ -30,6 +31,18 @@ class NonIDF_Properties(object):
           "information" : (4,lambda msg:   logger.information(msg)),
           "debug" :       (5,lambda msg:   logger.debug(msg))}
 
+    # The default location of the archive load log file
+    if platform == "linux" or platform == "linux2":
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:35]:
```suggestion
    if platform.startswith("linux"):
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 3]
```diff
@@ -178,10 +192,40 @@ def mapmask_ref_ws(self):
     def mapmask_ref_ws(self,val):
         object.__setattr__(self,'_mapmask_ref_ws',val)
 
+    # -----------------------------------------------------------------------------
+    @property
+    def arhive_upload_log_file(self):
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:197]:
```suggestion
    def archive_upload_log_file(self):
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 4]
```diff
@@ -30,6 +31,18 @@ class NonIDF_Properties(object):
           "information" : (4,lambda msg:   logger.information(msg)),
           "debug" :       (5,lambda msg:   logger.debug(msg))}
 
+    # The default location of the archive load log file
+    if platform == "linux" or platform == "linux2":
+        # linux default log file location
+        arhive_upload_log_template = '/archive/NDX{0}/Instrument/logs/lastrun.txt'
+    elif platform == "win32":
+        # windows default log file location
+        arhive_upload_log_template = r'\\isis\inst$\NDX{0}\Instrument\logs\lastrun.txt'
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:40]:
```suggestion
        archive_upload_log_template = r'\\isis\inst$\NDX{0}\Instrument\logs\lastrun.txt'
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 5]
```diff
@@ -30,6 +31,18 @@ class NonIDF_Properties(object):
...
+    else:
+        arhive_upload_log_template = ''
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:42]:
```suggestion
        archive_upload_log_template = ''
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 6]
```diff
@@ -178,10 +192,40 @@ def mapmask_ref_ws(self):
...
+            if len(NonIDF_Properties.arhive_upload_log_template)>0:
+                trial_file = NonIDF_Properties.arhive_upload_log_template.format(self.instr_name)
+                self._set_arcive_update_log(trial_file,False)
+            else:
+                object.__setattr__(self,'_archive_upload_log_file','')
+
+        return self._archive_upload_log_file
+
+    @arhive_upload_log_file.setter
+    def arhive_upload_log_file(self,filename):
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:213]:
```suggestion
    def archive_upload_log_file(self,filename):
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 7]
```diff
@@ -178,10 +192,40 @@ def mapmask_ref_ws(self):
...
+        if self._archive_upload_log_file is None:
+            if len(NonIDF_Properties.arhive_upload_log_template)>0:
+                trial_file = NonIDF_Properties.arhive_upload_log_template.format(self.instr_name)
+                self._set_arcive_update_log(trial_file,False)
+            else:
+                object.__setattr__(self,'_archive_upload_log_file','')
+
+        return self._archive_upload_log_file
+
+    @arhive_upload_log_file.setter
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:212]:
```suggestion
    @archive_upload_log_file.setter
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 8]
```diff
@@ -178,10 +192,40 @@ def mapmask_ref_ws(self):
...
+    def arhive_upload_log_file(self,filename):
+        self._set_arcive_update_log(filename)
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:214]:
```suggestion
        self._set_archive_update_log(filename)
```

File: scripts/Inelastic/Direct/NonIDF_Properties.py [hunk 9]
```diff
@@ -178,10 +192,40 @@ def mapmask_ref_ws(self):
...
+
+    def _set_arcive_update_log(self,filename,report_failure = True):
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/NonIDF_Properties.py:216]:
```suggestion
    def _set_archive_update_log(self,filename,report_failure = True):
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 10]
```diff
@@ -95,6 +95,10 @@ def __init__(self, instrumentName, web_var=None):
         web_vars = self._wvs.get_all_vars()
         if web_vars:
             self.reducer.prop_man.set_input_parameters(**web_vars)
+        # if run on ISIS, information about the log files, responsible for
+        # storing data in archive
+        self._lat_commit_log_modification_time = None
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:100]:
```suggestion
        self._last_commit_log_modification_time = None
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 11]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+              returns: (True,run_number_written,'') if the run_number stored in the log is
+                      higher then the run number requested
+                      (False,run_number_written,'') if the stored number is lower then the requested
+
+            If progress log is nod defined or not available, the method returns True, last known run number
+            and additional text information indicating the reason for failure
+            so further checks are necessary to verify if actual file is indeed available
+        """
+        propman = self.reducer.prop_man
+        if len(propman.arhive_upload_log_file)==0 :
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:466]:
```suggestion
        if len(propman.archive_upload_log_file)==0 :
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 12]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+            return (True,0,'log test disabled as no log file available')
+
+        mod_time = os.path.getmtime(propman.arhive_upload_log_file)
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:469]:
```suggestion
        mod_time = os.path.getmtime(propman.archive_upload_log_file)
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 13]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+        if self._lat_commit_log_modification_time == mod_time: # Still old data in archive
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:470]:
```suggestion
        if self._last_commit_log_modification_time == mod_time: # Still old data in archive
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 14]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+            run_num = self._last_runnum_added_to_archive
+            return (run_num >= run_number_requested,run_num,'no new data have been added to archive')
+        self._lat_commit_log_modification_time = mod_time
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:473]:
```suggestion
        self._last_commit_log_modification_time = mod_time
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 15]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+        try:
+            with open(propman.arhive_upload_log_file) as fh:
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:475]:
```suggestion
            with open(propman.archive_upload_log_file) as fh:
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 16]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+                contents = fh.read()
+        except:
+            return(False,self._last_runnum_added_to_archive,
+                   'Error accessing log file {0}'.format(propman.arhive_upload_log_file))
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:479]:
```suggestion
                   'Error accessing log file {0}'.format(propman.archive_upload_log_file))
```

File: scripts/Inelastic/Direct/ReductionWrapper.py [hunk 17]
```diff
@@ -443,7 +447,43 @@ def _run_pause(self, timeToWait=0):
...
+
+        mod_time = os.path.getmtime(propman.arhive_upload_log_file)
+        if self._lat_commit_log_modification_time == mod_time: # Still old data in archive
+            run_num = self._last_runnum_added_to_archive
+            return (run_num >= run_number_requested,run_num,'no new data have been added to archive')
+        self._lat_commit_log_modification_time = mod_time
+        try:
+            with open(propman.arhive_upload_log_file) as fh:
+                contents = fh.read()
+        except:
```
Comment (martyngigg) [inline scripts/Inelastic/Direct/ReductionWrapper.py:477]: I think here we can just catch `OSError`?

Catching everything also picks up things like `KeyboardInterrrupt` that you don't want to catch.

File: scripts/test/DirectPropertyManagerTest.py [hunk 18]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
         self.assertEqual(got.material, cyl.material)
         self.assertEqual(got.shape, cyl.shape)
 
+    #
+    def test_lastrun_log_default(self):
+        #
+        if platform == "linux" or platform == "linux2":
+            self.assertEqual(PropertyManager.arhive_upload_log_template,
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1266]:
```suggestion
            self.assertEqual(PropertyManager.archive_upload_log_template,
```

File: scripts/test/DirectPropertyManagerTest.py [hunk 19]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
...
+                             '/archive/NDX{0}/Instrument/logs/lastrun.txt')
+            log_dir = '/archive/NDXMARI/Instrument/logs/'
+        elif platform == "darwin":
+            self.assertEqual(PropertyManager.arhive_upload_log_template,'')
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1270]:
```suggestion
            self.assertEqual(PropertyManager.archive_upload_log_template,'')
```

File: scripts/test/DirectPropertyManagerTest.py [hunk 20]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
...
+            log_dir = ''
+        elif platform == "win32":
+            self.assertEqual(PropertyManager.arhive_upload_log_template,
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1273]:
```suggestion
            self.assertEqual(PropertyManager.archive_upload_log_template,
```

File: scripts/test/DirectPropertyManagerTest.py [hunk 21]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
...
+                             r'\\isis\inst$\NDX{0}\Instrument\logs\lastrun.txt')
+            log_dir = r'\\isis\inst$\NDXMARI\Instrument\logs\\'
+
+        propman = self.prop_man
+
+        if os.path.isdir(log_dir):
+            # in case test server or test machine is connected to the archive
+            self.assertEqual(propman.arhive_upload_log_file,
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1281]:
```suggestion
            self.assertEqual(propman.archive_upload_log_file,
```

File: scripts/test/DirectPropertyManagerTest.py [hunk 22]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
...
+                             os.path.normpath(log_dir+'lastrun.txt'))
+        else:
+            self.assertEqual(propman.arhive_upload_log_file,'')
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1284]:
```suggestion
            self.assertEqual(propman.archive_upload_log_file,'')
```

File: scripts/test/DirectPropertyManagerTest.py [hunk 23]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
...
+
+        test_dir = config.getString('defaultsave.directory')
+        test_file = os.path.normpath(test_dir + 'lastrun.txt')
+        f = open(test_file, 'w')
+        f.write('aaaaaa')
+        f.close()
+
+        propman.arhive_upload_log_file = test_file
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1292]:
```suggestion
        propman.archive_upload_log_file = test_file
```

File: scripts/test/DirectPropertyManagerTest.py [hunk 24]
```diff
@@ -1259,8 +1259,43 @@ def test_abs_shapes_container(self):
...
+        self.assertEqual(propman.arhive_upload_log_file,test_file)
```
Comment (martyngigg) [inline scripts/test/DirectPropertyManagerTest.py:1293]:
```suggestion
        self.assertEqual(propman.archive_upload_log_file,test_file)
```

File: scripts/test/ReductionWrapperTest.py [hunk 25]
```diff
@@ -288,6 +288,64 @@ def test_return_run_list(self):
         self.assertTrue('SR_MAR000400#1_reduced' in mtd)
         self.assertTrue('SR_MAR000400#2_reduced' in mtd)
 
+    def test_check_archive_logs(self):
```
Comment (martyngigg) [inline scripts/test/ReductionWrapperTest.py:291]: Good that we have extra tests 👍

**Review by martyngigg — CHANGES_REQUESTED** [review]
The code improves the handling of last finished run.

There are some typos that I expect have been propagated by an IDE autocomplete to deal with.

File: scripts/Inelastic/Direct/ReductionWrapper.py (hunk 17 above)
Comment (abuts) [reply scripts/Inelastic/Direct/ReductionWrapper.py:477]: I have carefully looked into `OSError` and it is insufficient type of error to catch as `IOError` is different and also can be risen here. The issue comes from the fact that the log file can be suddenly modified by archive committing process, so the access may be denied or file destroyed or something else. The parent `EnvironmentError` may be suitable but the parent includes `KeyboardInterrupt`.  I think its reasonable to keep the code as it is and keyboard interrupt will be caught in other parts of the code, as it goes into waiting mode, if file access failed. We can of course try to catch `OSError` and `IOError` in sequence, but I am not sure this list is inclusive, bearing in mind any issue occurring with the file.

Comment (abuts) [reply scripts/Inelastic/Direct/ReductionWrapper.py:477]: This is the most uncertain part of the code with any errors difficult to reproduce and understand, so I better add additional exception check to the results analysis.

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Could someone merge this for Excitations as they need it for next cycle.

**Pull request merged**